File: Doubly Linked List/DLL.py
class Node:

    def __init__(self, val):
        self.prev = None
        self.val = val
        self.next = None

class DLL:

    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def push(self, val):
        node = Node(val)
        if self.head == None:
            self.head = node
            self.tail = self.head
        else:
            self.tail.next = node
            node.prev = self.tail
            self.tail = node
        self.length += 1
        return self

    def pop(self):
        if self.head == None:
            return None
        elif self.head.next == None and self.tail.prev == None:
            self.head = None
            self.tail = None
        else:
            oldTail = self.tail
            self.tail = self.tail.prev
            self.tail.next = None
            oldTail.prev = None
        self.length -= 1
        return self
          
    def shift(self):
        if self.length == 0: return None
        elif self.length == 1:
            self.length -= 1
            self.head = None
            self.tail = None
        else:
            oldHead = self.head
            self.head = oldHead.next
            oldHead.next = None
            self.head.prev = None
            self.length -= 1
        return self

    def unshift(self, val):
        newNode = Node(val)
        if(self.head==None and self.tail==None): self.head = self.tail = newNode
        else:
            tempHead=self.head
            newNode.next=tempHead
            self.head.prev=newNode
            self.head=newNode
        self.length += 1
        return self

    def get(self, index):
        if(self.length <= 0 or index >= self.length or index < 0): return None
        mid = self.length // 2
        if index <= mid:
            temp = self.head
            for i in range(0, index+1, +1):
                if i == index: return temp
                temp = temp.next 
        elif index > mid:
            temp = self.tail
            for i in range(self.length-1, index-1, -1):
                if i == index: return temp
                temp = temp.prev

    def set(self, index, val):
        node = self.get(index)
        node.val = val
        return self

    def insert(self, index, val):
        if index < 0 or index > self.length: return None
        elif index == 0: self.unshift(val)
        elif index == self.length: self.push(val)
        else:
            node = Node(val)
            oldPrevNode = self.get(index-1)
            oldNextNode = oldPrevNode.next
            node.prev = oldPrevNode
            node.next = oldPrevNode.next
            oldPrevNode.next = node
            oldNextNode.prev = node
            self.length += 1
        return self 

    def remove(self, index):
        if index < 0 or index >= self.length: return None
        elif index == 0: return self.shift()
        elif index == self.length-1: return self.pop()
        else:
            prevNode = self.get(index-1)
            remNode = prevNode.next
            prevNode.next = remNode.next
            remNode.prev = None
            remNode.next = None
            self.length -= 1


    def traverse(self):
        if self.length == 0: return None
        else:
            current = self.head
            while(current):
                print(current.val)
                current = current.next
        

dll = DLL()

# --------------- Push test cases ---------------
# dll.push(3)
# dll.push(4)
# dll.push(5)
# print(dll.tail.prev.prev.val)

# --------------- Pop test cases ---------------
# dll.push(3)
# dll.push(4)
# dll.push(5)
# print(dll.pop())

# --------------- Shift test cases ---------------
# dll.shift()
# dll.shift()
# dll.shift()
# print(dll.head)

# --------------- UnShift test cases ---------------
# print(dll.unshift(5).head.val)
# print(dll.unshift(6).head.val)
# print(dll.unshift(7).head.val)
# print(dll.length)

# --------------- Get test cases ---------------
# dll.push("hi")
# dll.push("there")
# dll.push("omi")
# dll.push("3")
# dll.push("4")
# dll.push("5")
# dll.push("6")
# print(dll.get(0))

# --------------- Set test cases ---------------
# dll.push("hi")
# dll.push("there")
# dll.push("omi")
# print("previously:", dll.get(2).val)
# dll.set(2, "!")
# print("later:", dll.get(2).val)

# --------------- Insert test cases ---------------
# dll.push("A")
# dll.push("B")
# dll.push("C")
# dll.push("E")
# print(dll.length)
# dll.insert(3, "D")
# print(dll.length)
# dll.traverse()

# --------------- Remove test cases ---------------
# dll.push("A")
# dll.push("B")
# dll.push("C")
# dll.push("D")
# dll.traverse()
# print("------------------------------------------------")
# dll.remove(1)
# dll.traverse()

File: Doubly Linked List/DLL.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
    this.prev = null;
  }
}

class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }
  push(val) {
    var node = new Node(val);
    if (this.head === null) {
      this.head = node;
      this.tail = node;
    } else {
      this.tail.next = node;
      node.prev = this.tail;
      this.tail = node;
    }
    this.length++;
    return this;
  }
  unshift(val) {
    var node = new Node(val);
    if (this.head === null && this.tail === null) {
      this.head = node;
      this.tail = node;
    } else {
      var tempHead = this.head;
      node.next = tempHead;
      this.head.prev = node;
      this.head = node;
    }
    this.length++;
    return this;
  }
  shift() {
    if (this.length === 0) return undefined;
    var oldHead = this.head;
    if (this.length === 1) {
      this.head === null;
      this.tail === null;
    } else {
      this.head = oldHead.next;
      oldHead.next = null;
      this.head.prev = null;
    }
    this.length--;
    return oldHead;
  }
  pop() {
    return undefined;
  }
  get(index) {
    if (this.length <= 0 || index >= this.length || index < 0) {
      return null;
    }
    var mid = Math.floor(this.length / 2);
    if (index <= mid) {
      var current = this.head;
      for (let i = 0; i < index + 1; i++) {
        if (i === index) {
          return current;
        }
        current = current.next;
      }
    } else if (index > mid) {
      var current = this.tail;
      for (let i = this.length - 1; index - 1; i--) {
        if (i === index) {
          return current;
        }
        current = current.prev;
      }
    }
  }
  set(index, value) {
    var node = this.get(index);
    if (node === false) {
      return false;
    }
    node.val = value;
    return true;
  }

  insert(index, val) {
    if (index < 0 || index >= this.length) {
      return false;
    } else if (index === 0) {
      return this.unshift(val);
    } else if (index === this.length - 1) {
      return this.push(val);
    } else {
      var node = new Node(val);
      var prevNode = this.get(index - 1);
      node.next = prevNode.next;
      prevNode.next = node;
      node.prev = prevNode;
      node.next.prev = node;
      this.length++;
      return true;
    }
  }

  remove(index) {
    if (index < 0 || index >= this.length) {
      return undefined;
    } else if (index === 0) {
      return this.shift();
    } else if (index === this.length - 1) {
      return this.pop();
    } else {
      var prevNode = this.get(index - 1);
      var remNode = prevNode.next;
      prevNode.next = remNode.next;
      remNode.prev = null;
      remNode.next = null;
      this.length--;
      return remNode;
    }
  }
  pop() {
    if (this.length === 0) {
      return undefined;
    } else if (this.length === 1) {
      this.length--;
      return this.head;
    } else {
      var oldTail = this.tail;
      this.tail = oldTail.prev;
      this.tail.next = null;
      oldTail.next = null;
      oldTail.prev = null;
      this.length--;
      return oldTail;
    }
  }
  reverse() {}
}

// ------------------------------- Test cases
let doublyLinkedList = new DoublyLinkedList();
doublyLinkedList.push(5).push(10).push(15).push(20);
doublyLinkedList.reverse(); // singlyLinkedList;
doublyLinkedList.length; // 4
doublyLinkedList.head.val; // 20
doublyLinkedList.head.next.val; // 15
doublyLinkedList.head.next.next.val; // 10
doublyLinkedList.head.next.next.next.val; // 5
